### PR TITLE
Parse text literal, lambda and multi-parameter function

### PR DESCRIPTION
**Pull request by JaroslavTulach**
### Pull Request Description

Construct `IR` for text literal, lambda and multi-parameter function using the new Rust parser.

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide build` and `./run ide watch`.

[ci no changelog needed]

### CODE REVIEW

**Comment by JaroslavTulach** [comment]
This PR extracts and integrates few more conversions from #3611 to `develop` branch.

File: engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java
Comment (hubertp) [inline engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java:701]: So it is ignoring `"`?

File: engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java
Comment (hubertp) [inline engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java:707]: that looks spurious?

File: engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java
Comment (hubertp) [inline engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java]:
```suggestion
        throw new UnhandledEntity(tree, "translateComment");
```

File: engine/runtime/src/test/java/org/enso/compiler/EnsoCompilerTest.java
Comment (hubertp) [inline engine/runtime/src/test/java/org/enso/compiler/EnsoCompilerTest.java]: I guess the real name of the test is sth like like `testMultiParameterFunction` ?

**Review by hubertp — APPROVED** [review]
nits

File: engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java:701]: Yes. I'd like this to be handled by the parser itself, but that is not the case yet. I'd like the full text literal to be provided to me by the parser, but @kazcw explained to me that's not possible - I can only get it line by line. Still quoting, character conversions, etc. would better be handled by the parser itself.

File: engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java:707]: It hides the `head` variable from the rest of the scope.

File: engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java
Comment (kazcw) [reply engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java:701]: > I'd like the full text literal to be provided to me by the parser, but @kazcw explained to me that's not possible

Right, the problem is not well-defined. A `TextLiteral` that can contain splices cannot in general be represented as a `String`. I think splices are new (it looks like they aren't yet representable in the IR?), so from the perspective of implementing `TreeToIr` right now this representation might look overly-complicated; but if the parser provided an interface producing a `String` that only worked when the input didn't contain splices, we'd have no use for it once we had a general implementation that handles any `TextLiteral` as a sequence of text sections and expressions. Your labor tying strings together now is the foundation of a new language feature.

> Still quoting, character conversions, etc. would better be handled by the parser itself.

They are. `codeRepr` is not the parsed text, it stringifies back to the literal source code... The parser handles escapes, splices, quotes, and the block text left-trim algorithm, and stores the result in the `TextLiteral`'s fields. Call `getElements` and iterate to use the *parsed* data. Please hold off on implementing that until #3716 is merged though, or work from that branch; I've made improvements to text parsing there.

File: engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java
Comment (kazcw) [inline engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java]: An [argument block](https://github.com/enso-org/design/blob/wip/wd/enso-spec/epics/enso-spec-1.0/03.%20Code%20format%20and%20layout.md#argument-list-code-blocks) applies all of its arguments (`getArguments`) to its lhs (`getLhs`). It is not a type of comment...

File: engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java
Comment (kazcw) [inline engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java:1753]: Not a comment, but in some cases comments produce a parse containing `ArgumentBlockApplication`s, so I see how this happened. All source code, including non-semantic text like whitespace and non-doc-comments, has to be represented in the `Tree` types. `ArgumentBlockApplication` is the default block type--any increase in indentation that doesn't fit the rules for an operator block or a body block is an argument block; so to represent some source, you'll get an `ArgumentBlockApplication` that doesn't contain any expressions. This is only the way that *plain* comments are represented. As I've mentioned, the parser doesn't yet handle doc comments; it treats all comments as plain comments. Doc comments, once implemented, will have their own node and token types.

I've postponed the doc comment implementation (https://www.pivotaltracker.com/story/show/182497173 and https://www.pivotaltracker.com/story/show/182497187) until after integration, because the plan is to lex/parse markdown, and that is not needed for compatibility with the current parser. However, I missed an intermediate step: For compatibility with the current parser, we do need to distinguish doc comments and place them in the AST. I'll implement that soon: https://www.pivotaltracker.com/story/show/183417662. Until then, doc comments don't exist in the lexer, so it's not necessary to try to extract them from the AST. Once they're represented, the data will be much easier to use.

**Review by kazcw — DISMISSED** [review]

**Comment by kazcw** [comment]
Please, if something seems wrong, don't implement it; let's talk about it. Interpreting a `Tree`'s `codeRepr` in the backend is never correct--if that's ever actually necessary to produce the right IR, that's a bug in the parser.

File: engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java:701]: OK, I got it. I am supposed to use `getElements`...

> hold off on implementing that until https://github.com/enso-org/enso/pull/3716 is merged though

...once #3716 is integrated. Meanwhile keeping this `fullTxt.substring(1, fullTxt.length() - 1)` hack is enough.

File: engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java]: Right now about fifty files in our standard library fail to pass thru `TreeToIr` conversion because documentation comment is represented as `Tree.ArgumentBlockApplication` - I have to process it somehow. It would help if [pivotal 183417662](https://www.pivotaltracker.com/n/projects/2539304/stories/183417662) was implemented and:
```
    ## Measure the amount of time it takes to execute a given computation.
    
       Arguments:
       - act: The action to perform.
       - label: A name for the measurement.
       - iter_size: The number of runs per iteration.
       - num_iters: The number of iterations per measurement.
    
       > Example
         Measure a computation called "foo" with an iteration size of 2 and a number
         of iterations of 1.
    
             import Standard.Examples
             import Standard.Test.Bench
    
             example_measure =
                 Bench.measure Examples.get_boolean "foo" iter_size=2 num_iters=1
    measure : Any -> Text -> Integer -> Integer -> Nothing
    measure = ~act -> label -> iter_size -> num_iters ->
        42
```
wasn't represented as `Tree.ArgumentBlockApplication` - but it is. Until the Rust part delivers better result for documentation comments, converting `Tree.ArgumentBlockApplication` by `translateComment` does the least harm.

File: engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java:1753]: Comment processing removed in [466ec48](https://github.com/enso-org/enso/pull/3747/commits/466ec486388ad45adecaad6f6278bf80f84bc821).

**Comment by JaroslavTulach** [comment]
> Please, if something seems wrong, don't implement it; let's talk about it

I am not dare to decide what is _right or wrong_. I can only decide whether the `TreeToIr` output matches the `IR`  created by `AstToIr` or not. When it doesn't I am trying to fix that by any available means.

File: engine/runtime/src/test/java/org/enso/compiler/EnsoCompilerTest.java
Comment (JaroslavTulach) [reply engine/runtime/src/test/java/org/enso/compiler/EnsoCompilerTest.java]: Done in [d13cc31](https://github.com/enso-org/enso/pull/3747/commits/d13cc31dfdbb3bfa40a864baafb033a762655a5c)

File: engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java]: I removed this diff in [466ec48](https://github.com/enso-org/enso/pull/3747/commits/466ec486388ad45adecaad6f6278bf80f84bc821)

**Pull request merged**
